docs/src/proposals/off-chain-message-signing.md

```diff
@@ -0,0 +1,78 @@
+# Off-chain message signing
+## Motivation
+
+There is ecosystem demand for a method of signing non-transaction messages with
+a Solana wallet. Typically this is some kind of "proof of wallet ownership" for
+entry into a whitelisted system.
+
+Some inspiration can be gleaned from relevant portions of Ethereum's
+[EIP-721](https://eips.ethereum.org/EIPS/eip-712)
+
+## Considerations
+
+* Security
+  * Off-chain message signatures should not be valid transaction message signatures
+* Future-proofing
+  * Versioning
+  * Co-exist with versioned transaction messages and extended length transactions
+* Compatibility
+  * Hardware wallet signing
+  * Localization
+
+## Payload format
+
+| 0x00  ..= 0x0f | 0x10           | 0x11           | 0x12  ..= 0x13 | 0x14 .. |
+| :------------: | :------------: | :------------: | :------------: | :-----: |
+| Signing Domain | Header Version | Message Format | Message Length | Message |
+
+The payload consists of a 16-byte signing domain specifier, 4-byte header and
+n-byte message.
+
+### Signing Domain Specifier
+The signing domain specifier is a prefix byte string used to give similar structure
+to all off-chain message signatures. We assign its value to be:
+```
+b"\xffsolana offchain"
```

Comment:
Since the off-chain message signing scheme is targeted at a wide variety of unrelated apps, we might want to force off-chain apps to include their own application identifiers (analogous to ALPN in TLS).

A static signing domain only prevents confusion between on-chain and off-chain messaging. Apps will want separate off-chain signing domains regardless. I propose adding another 16 bytes that the app can choose.

Reply:
For format 0 and 1 messages, a fixed field would eat into the (limited) size available for messages. Perhaps a non-normative recommendation to use N bytes at the start of the message to specify app-specific sub-domain?

Reply:
Signing domains only work if they are always enforced. Otherwise an app that opts-out can forge a message colliding with all other apps which would not be obvious to the user.
Could you elaborate what limits message length? (hardware wallets I think?)

Reply:
Yep, per the message format section, formats 0 and 1 have a limited message length, for support by hardware wallets.

Re: signing domains:
> I propose adding another 16 bytes that the app can choose.

If apps get to choose this, then they can just lie about their domain in any case, right? It's not an area where I have particularly deep expertise, so I will absolutely defer to those more knowledgeable on this, I'm just not clear on how putting this in the header (as a fixed field) vs in the body (as an optional part of the payload, that apps that care about it's presence can check for) differs.

Reply:
> If apps get to choose this, then they can just lie about their domain in any case, right? 

Good point – the hardware wallet could prompt the user to confirm the app ID that's requesting a signature, similar to how on-chain transfers signing requests prompt for the destination address and amount.

Reply:
Gotcha - I understand now; a mandatory field presents opportunities for the signer to perform additional validation activities

Reply:
thoughts on extending the app domain specifier to 32-bytes, so it can support a common pubkey/address?  I know it's more "eating into the message", but this feature is for auth, not writing home to mum.

Reply:
"16 bytes ought to be enough for anybody"

Reply:
But wouldn't it need to be 33 bytes, with 0xff prefix?

Reply:
> But wouldn't it need to be 33 bytes, with 0xff prefix?

Nah, only the global domain specifier needs that crafted marker ('cause it's the first byte of the payload). App domain specifier can be whatever